ceno_zkvm/src/instructions/riscv/mulh.rs

```diff
@@ -7,195 +87,170 @@ use crate::{
     circuit_builder::CircuitBuilder,
     error::ZKVMError,
     expression::Expression,
-    gadgets::SignedExtendConfig,
+    gadgets::{IsEqualConfig, SignedExtendConfig},
     instructions::{
         Instruction,
         riscv::{
             RIVInstruction,
-            constants::{BIT_WIDTH, UInt, UIntMul},
+            constants::{BIT_WIDTH, UInt},
             r_insn::RInstructionConfig,
         },
     },
     uint::Value,
+    utils::i64_to_base,
     witness::LkMultiplicity,
 };
 use core::mem::MaybeUninit;
 
-/// This config handles R-Instructions that represent registers values as 2 * u16.
-#[derive(Debug)]
-pub struct ArithConfig<E: ExtensionField> {
-    r_insn: RInstructionConfig<E>,
+pub struct MulhInstructionBase<E, I>(PhantomData<(E, I)>);
 
-    rs1_read: UInt<E>,
-    rs2_read: UInt<E>,
-    rd_written: UIntMul<E>,
+pub struct MulhOp;
+impl RIVInstruction for MulhOp {
+    const INST_KIND: InsnKind = InsnKind::MULH;
 }
-
-pub struct MulhInstructionBase<E, I>(PhantomData<(E, I)>);
+pub type MulhInstruction<E> = MulhInstructionBase<E, MulhOp>;
 
 pub struct MulhuOp;
 impl RIVInstruction for MulhuOp {
     const INST_KIND: InsnKind = InsnKind::MULHU;
 }
 pub type MulhuInstruction<E> = MulhInstructionBase<E, MulhuOp>;
 
+pub struct MulhsuOp;
+impl RIVInstruction for MulhsuOp {
+    const INST_KIND: InsnKind = InsnKind::MULHSU;
+}
+pub type MulhsuInstruction<E> = MulhInstructionBase<E, MulhsuOp>;
+
+pub struct MulhConfig<E: ExtensionField> {
+    rs1_read: UInt<E>,
+    rs2_read: UInt<E>,
+    rd_written: UInt<E>,
+    sign_deps: MulhSignDependencies<E>,
+    prod_low: UInt<E>,
+    r_insn: RInstructionConfig<E>,
+}
+
+enum MulhSignDependencies<E: ExtensionField> {
```

Comment:
This is a good idea to make the declaration of `_Config` more succinct. I believe there are many opcodes can apply this pattern. What do you think, @hero78119 ?

Reply:
definitely it can clean up on opcode which expose `Option<>` as internal field, so we can control them clearly in enum
@KimiWu123 if you interesting in the cleanup it's good to have an issue to track it :)